File: .claude/hooks/load-project-skills.py
#!/usr/bin/env python
"""Auto-load project skills when editing relevant files.

Injects each skill at most once per session via hookSpecificOutput.additionalContext.
Triggered as a PreToolUse hook on Edit|Write.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path

SKILL_MAP = [
    ("/app/backend/", "structuring-kotlin-ddd-code"),
    ("/app/frontend/", "frontend-storybook-architecture"),
    ("/.agents/skills/", "managing-skill-evolution"),
    ("/infra/env/", "env-sync"),
]


def main() -> int:
    try:
        payload = json.load(sys.stdin)
    except Exception:
        return 0

    file_path = (payload.get("tool_input") or {}).get("file_path") or ""
    session_id = payload.get("session_id") or "default"
    if not file_path:
        return 0

    norm = file_path.replace("\\", "/")
    skill_dir = next((name for token, name in SKILL_MAP if token in norm), None)
    if not skill_dir:
        return 0

    repo_root = Path(__file__).resolve().parents[2]
    skill_file = repo_root / ".agents" / "skills" / skill_dir / "SKILL.md"
    if not skill_file.is_file():
        return 0

    sentinel_dir = repo_root / ".claude" / "cache" / "skill-loaded" / session_id
    sentinel_dir.mkdir(parents=True, exist_ok=True)
    sentinel = sentinel_dir / skill_dir
    if sentinel.exists():
        return 0
    sentinel.touch()

    content = skill_file.read_text(encoding="utf-8")
    rel = f".agents/skills/{skill_dir}/SKILL.md"
    out = {
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "additionalContext": (
                f"Auto-loaded project skill from {rel} (edit target matched its scope):\n\n"
                f"{content}"
            ),
        }
    }
    json.dump(out, sys.stdout)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .claude/hooks/load-project-skills.js
#!/usr/bin/env node
// PreToolUse hook — auto-load project skills when editing relevant files.
// Injects each skill at most once per session via hookSpecificOutput.additionalContext.
// Triggered as a PreToolUse hook on Edit|Write.
//
// Cross-platform (Node.js): works identically on Windows Git Bash / macOS / Linux
// without depending on Python or jq.

const fs = require('fs');
const path = require('path');

const SKILL_MAP = [
  ['/app/backend/',     'structuring-kotlin-ddd-code'],
  ['/app/frontend/',    'frontend-storybook-architecture'],
  ['/.agents/skills/',  'managing-skill-evolution'],
  ['/infra/env/',       'env-sync'],
];

function main() {
  // Read entire stdin (Claude Code provides a JSON blob).
  let raw;
  try {
    raw = fs.readFileSync(0, 'utf8');
  } catch {
    return 0;
  }

  let payload;
  try {
    payload = JSON.parse(raw);
  } catch {
    return 0;
  }

  const filePath = (payload.tool_input && payload.tool_input.file_path) || '';
  const sessionId = payload.session_id || 'default';
  if (!filePath) return 0;

  const norm = filePath.replace(/\\/g, '/');
  const match = SKILL_MAP.find(([token]) => norm.includes(token));
  if (!match) return 0;
  const skillDir = match[1];

  // repo_root = <hookFile>/../..  (= .claude/hooks/ -> .claude/ -> repo root)
  const repoRoot = path.resolve(__dirname, '..', '..');
  const skillFile = path.join(repoRoot, '.agents', 'skills', skillDir, 'SKILL.md');
  if (!fs.existsSync(skillFile) || !fs.statSync(skillFile).isFile()) return 0;

  // Sentinel — one injection per (session_id, skill) pair.
  const sentinelDir = path.join(repoRoot, '.claude', 'cache', 'skill-loaded', sessionId);
  fs.mkdirSync(sentinelDir, { recursive: true });
  const sentinel = path.join(sentinelDir, skillDir);
  if (fs.existsSync(sentinel)) return 0;
  fs.writeFileSync(sentinel, '');

  const content = fs.readFileSync(skillFile, 'utf8');
  const rel = `.agents/skills/${skillDir}/SKILL.md`;
  const out = {
    hookSpecificOutput: {
      hookEventName: 'PreToolUse',
      additionalContext:
        `Auto-loaded project skill from ${rel} (edit target matched its scope):\n\n` +
        content,
    },
  };
  process.stdout.write(JSON.stringify(out));
  return 0;
}

process.exit(main());
